assembly/Quat.ts

```diff
@@ -0,0 +1,388 @@
+                    this.w = (m12 - m21) * rs;
+                    this.y = (m01 + m10) * rs;
+                    this.z = (m02 + m20) * rs;
+                } else {
+                    // ZDiagDomMatrix
+                    rs = (m22 - (m00 + m11)) + 1.0;
+                    rs = Mathf.sqrt(rs);
+
+                    this.z = rs * 0.5;
+                    rs = 0.5 / rs;
+                    this.w = (m01 - m10) * rs;
+                    this.x = (m20 + m02) * rs;
+                    this.y = (m21 + m12) * rs;
+                }
+            } else if (m11 > m22) {
+                // YDiagDomMatrix
+                rs = (m11 - (m22 + m00)) + 1.0;
+                rs = Mathf.sqrt(rs);
+
+                this.y = rs * 0.5;
+                rs = 0.5 / rs;
+                this.w = (m20 - m02) * rs;
+                this.z = (m12 + m21) * rs;
+                this.x = (m10 + m01) * rs;
+            } else {
+                // ZDiagDomMatrix
+                rs = (m22 - (m00 + m11)) + 1.0;
+                rs = Mathf.sqrt(rs);
+
+                this.z = rs * 0.5;
+                rs = 0.5 / rs;
+                this.w = (m01 - m10) * rs;
+                this.x = (m20 + m02) * rs;
+                this.y = (m21 + m12) * rs;
+            }
+        }
+
+        return this;
+    }
+
+    slerp(lhs: Quat, rhs: Quat, alpha: f32): Quat {
+        // Algorithm sourced from:
+        // http://www.euclideanspace.com/maths/algebra/realNormedAlgebra/quaternions/slerp/
+
+        var lx = lhs.x;
+        var ly = lhs.y;
+        var lz = lhs.z;
+        var lw = lhs.w;
+        var rx = rhs.x;
+        var ry = rhs.y;
+        var rz = rhs.z;
+        var rw = rhs.w;
+
+        // Calculate angle between them.
+        var cosHalfTheta = lw * rw + lx * rx + ly * ry + lz * rz;
+
+        if (cosHalfTheta < 0.0) {
+            rw = -rw;
+            rx = -rx;
+            ry = -ry;
+            rz = -rz;
+            cosHalfTheta = -cosHalfTheta;
+        }
+
+        // If lhs == rhs or lhs == -rhs then theta == 0 and we can return lhs
+        if (Mathf.abs(cosHalfTheta) >= 1.0) {
+            this.w = lw;
+            this.x = lx;
+            this.y = ly;
+            this.z = lz;
+            return this;
+        }
+
+        // Calculate temporary values.
+        var halfTheta = Mathf.acos(cosHalfTheta);
+        var sinHalfTheta = Mathf.sqrt(1.0 - cosHalfTheta * cosHalfTheta);
+
+        // If theta = 180 degrees then result is not fully defined
+        // we could rotate around any axis normal to qa or qb
+        if (Mathf.abs(sinHalfTheta) < 0.001) {
+            this.w = (lw * 0.5 + rw * 0.5);
+            this.x = (lx * 0.5 + rx * 0.5);
+            this.y = (ly * 0.5 + ry * 0.5);
+            this.z = (lz * 0.5 + rz * 0.5);
+            return this;
+        }
+
+        var ratioA = Mathf.sin((1.0 - alpha) * halfTheta) / sinHalfTheta;
+        var ratioB = Mathf.sin(alpha * halfTheta) / sinHalfTheta;
```

Comment:
```suggestion
        var angle = alpha * halfTheta;
        var ratioA = Mathf.sin(halfTheta - angle) / sinHalfTheta;
        var ratioB = Mathf.sin(angle) / sinHalfTheta;
```

Reply:
Would it make sense to multiply both with `1 / sinHalfTheta` here aswell? Two divisions vs one division and two multiplications 😅

Reply:
> Would it make sense to multiply both with 1 / sinHalfTheta here aswell? 

No, usually two float point divisions better